File: node/src/rpc/lb.ts
import type {
	MethodDescriptor,
	ServiceInstanceInfo,
} from "../pb/servicebridge/v1/registry";
import type { CircuitBreakerRegistry } from "./circuit-breaker";

// Power-of-Two-Choices load balancer with per-pod inflight tracking (ADR 0001).
// Local CB filter excludes OPEN instances; runtime `is_unhealthy_since_unix_ms` hint
// excludes instances marked unhealthy by the runtime within the last
// HEALTH_HINT_TTL_MS. After the TTL the hint is treated as stale (runtime may
// itself be broken) and the local CB carries the routing decision.
//
// @internal — см. ./README.md

// HEALTH_HINT_TTL_MS bounds how long a runtime-supplied unhealthy hint is
// trusted. Bound = 2× the runtime HealthTracker healing window (ADR 0001).
export const HEALTH_HINT_TTL_MS = 60_000;

export interface Candidate {
	descriptor: MethodDescriptor;
	instance: ServiceInstanceInfo;
	// isUnhealthyAt is the wall-clock time the runtime first marked this
	// instance unhealthy through the watch snapshot (ADR 0001). Null when
	// the runtime has no opinion or has not yet wired the hint.
	isUnhealthyAt: Date | null;
}

export class NoLiveInstanceError extends Error {
	constructor(message?: string) {
		super(message ?? "rpc: no live instance");
		this.name = "NoLiveInstanceError";
	}
}

export class LoadBalancer {
	private inflight = new Map<string, number>();
	private now: () => number;
	private random: () => number;

	constructor(
		private readonly cb: CircuitBreakerRegistry,
		opts?: { now?: () => number; random?: () => number },
	) {
		this.now = opts?.now ?? Date.now;
		this.random = opts?.random ?? Math.random;
	}

	// pick selects an eligible candidate via Power of Two Choices. Eligibility:
	//   - candidate has a non-empty call_endpoint
	//   - the runtime health hint is stale or absent
	//   - the local circuit breaker is not OPEN, and — when HALF_OPEN — its single
	//     probe slot is free
	// Throws NoLiveInstanceError when no candidate is eligible. A tie on inflight
	// counters resolves to the first sampled candidate. Caller MUST pair every
	// successful pick with a `release` call in a finally block so inflight stays
	// balanced.
	//
	// HALF_OPEN single-probe: a HALF_OPEN instance is eligible only while its
	// probe slot is free (read-only `probeAvailable`). Once a winner is chosen,
	// pick CLAIMS the probe on that instance via `canCall`; the claim is released
	// by the matching cb.recordSuccess/recordFailure the caller runs after the
	// dispatch. pick runs to completion synchronously, so a concurrent caller's
	// later pick sees the claimed slot and routes elsewhere (or fails over) —
	// exactly one probe in flight per HALF_OPEN instance.
	//
	// P2C needs exactly two eligible candidates, so eligibility and sampling run
	// in a single pass with a two-slot reservoir. Materialising the eligible
	// subset would allocate an array on every call — with a few hundred callee
	// pods that is the dominant garbage the RPC path produces.
	pick(candidates: Candidate[]): Candidate {
		const now = this.now();
		let eligible = 0;
		let a: Candidate | undefined;
		let b: Candidate | undefined;
		for (const c of candidates) {
			if (!c.instance.callEndpoint) continue;
			const hintActive =
				c.isUnhealthyAt !== null &&
				now - c.isUnhealthyAt.getTime() < HEALTH_HINT_TTL_MS;
			if (hintActive) continue;
			if (!this.cb.probeAvailable(cbKey(c.instance))) continue;

			eligible++;
			if (eligible === 1) {
				a = c;
			} else if (eligible === 2) {
				b = c;
			} else {
				// Reservoir sampling, k=2: the n-th eligible candidate takes a
				// uniformly chosen slot with probability 2/n, which leaves every
				// eligible pair equally likely without holding the whole subset.
				const slot = Math.floor(this.random() * eligible);
				if (slot === 0) a = c;
				else if (slot === 1) b = c;
			}
		}
		if (a === undefined) throw new NoLiveInstanceError();
		let winner = a;
		if (b !== undefined) {
			// Reservoir sampling picks the pair uniformly but not its order: the
			// second eligible candidate can only ever land in slot B. Since a tie
			// on inflight resolves to slot A, an unshuffled reservoir would starve
			// that candidate on an idle fleet. One coin flip restores uniformity.
			winner =
				this.random() < 0.5 ? this.leastLoaded(a, b) : this.leastLoaded(b, a);
		}
		// Claim the HALF_OPEN probe slot on the winner only. CLOSED instances
		// always return true here without side effects.
		this.cb.canCall(cbKey(winner.instance));
		return winner;
	}

	private leastLoaded(a: Candidate, b: Candidate): Candidate {
		const ai = this.inflight.get(a.instance.instanceId) ?? 0;
		const bi = this.inflight.get(b.instance.instanceId) ?? 0;
		return ai <= bi ? a : b;
	}

	// acquire increments the inflight counter for an instance before the call
	// is dispatched. Always returns a matching release closure that decrements
	// the counter exactly once.
	acquire(instanceId: string): () => void {
		this.inflight.set(instanceId, (this.inflight.get(instanceId) ?? 0) + 1);
		let released = false;
		return () => {
			if (released) return;
			released = true;
			const cur = this.inflight.get(instanceId) ?? 0;
			if (cur <= 1) {
				this.inflight.delete(instanceId);
			} else {
				this.inflight.set(instanceId, cur - 1);
			}
		};
	}

	// inflightOf returns the current inflight count — for tests / metrics.
	inflightOf(instanceId: string): number {
		return this.inflight.get(instanceId) ?? 0;
	}
}

// keyCache memoises the derived breaker key per instance object. cbKey runs once
// per candidate per pick plus twice per completed call, and the template literal
// was allocating a short-lived string every time. Keying by the instance object
// keeps the cache self-clearing: InstanceCache rebuilds instance objects on each
// registry snapshot, so retired pods drop out with their objects and the cache
// cannot grow without bound the way a Map<instanceId, string> would.
const keyCache = new WeakMap<ServiceInstanceInfo, string>();

export function cbKey(instance: ServiceInstanceInfo): string {
	let key = keyCache.get(instance);
	if (key === undefined) {
		key = `${instance.serviceId}:${instance.instanceId}`;
		keyCache.set(instance, key);
	}
	return key;
}

File: go/internal/rpc/lb.go
package rpc

import (
	"math/rand/v2"
	"sync"
)

// DefaultHealthHintTTLMs bounds how long the runtime's unhealthy hint is
// trusted. It is twice the runtime's healing window: past that the hint is
// stale — a runtime that died mid-window would otherwise black-hole an instance
// forever — and the local breaker carries the decision alone.
const DefaultHealthHintTTLMs int64 = 60_000

// Candidate is one callee instance the registry offers for a call.
type Candidate struct {
	ServiceID   string
	ServiceName string
	InstanceID  string
	// Endpoint is the address peers dial for direct RPC. Empty means the
	// instance advertises none and is unreachable by either transport — the
	// runtime's resolver requires a non-empty call_endpoint too.
	Endpoint string
	// UnhealthySinceMs is the start of the runtime's current unhealthy window,
	// unix-ms. Zero means healthy or unknown.
	UnhealthySinceMs int64
}

// Key is the breaker and in-flight key for this candidate.
func (c Candidate) Key() BreakerKey {
	return BreakerKey{ServiceID: c.ServiceID, InstanceID: c.InstanceID}
}

// Healthy applies the runtime's hint with its trust window.
func Healthy(c Candidate, nowMs, hintTTLMs int64) bool {
	if c.UnhealthySinceMs == 0 {
		return true
	}
	if hintTTLMs <= 0 {
		hintTTLMs = DefaultHealthHintTTLMs
	}
	return nowMs-c.UnhealthySinceMs >= hintTTLMs
}

// PickStats explains what happened to a candidate set. The three counts are
// what separates the three ways selection fails, which have opposite fixes:
// nothing published this contract, the callee advertises no address, or the
// fleet is shedding.
type PickStats struct {
	// Total is every candidate the registry indexed for this contract.
	Total int
	// Addressed counts those advertising a call endpoint.
	Addressed int
	// Eligible counts those that also passed the health hint and the breaker.
	Eligible int
}

// Eligible filters cands and reports the counts behind the decision. It claims
// nothing and is safe to call for diagnostics.
func Eligible(cands []Candidate, allows func(Candidate) bool) ([]Candidate, PickStats) {
	stats := PickStats{Total: len(cands)}
	var out []Candidate
	for _, c := range cands {
		if c.Endpoint == "" {
			continue
		}
		stats.Addressed++
		if allows != nil && !allows(c) {
			continue
		}
		stats.Eligible++
		out = append(out, c)
	}
	return out, stats
}

// BalancerOption configures a Balancer at construction time.
type BalancerOption func(*Balancer)

// WithIntn injects the randomness source, a function returning a value in
// [0, n). Tests pin it to make the choice deterministic. It is only ever called
// while the balancer's lock is held, so it need not be safe for concurrent use.
func WithIntn(fn func(n int) int) BalancerOption {
	return func(b *Balancer) {
		if fn == nil {
			panic("rpc: WithIntn: nil source")
		}
		b.intn = fn
	}
}

// Balancer spreads calls over the eligible instances by power-of-two-choices on
// in-flight request count.
type Balancer struct {
	mu       sync.Mutex
	inflight map[BreakerKey]int
	intn     func(n int) int
}

// NewBalancer builds a balancer with process-global randomness.
func NewBalancer(opts ...BalancerOption) *Balancer {
	b := &Balancer{
		inflight: make(map[BreakerKey]int),
		intn:     rand.IntN,
	}
	for _, opt := range opts {
		opt(b)
	}
	return b
}

// Pick selects the instance to dispatch to and reserves an in-flight slot on
// it. The caller must Release the returned candidate once the call settles.
//
// allows is the side-effect-free eligibility check — health hint plus breaker.
// acquire claims the call slot on the winner alone, and may refuse: in the
// half-open state exactly one probe is allowed and a concurrent call may hold
// it. A refused winner is excluded and the draw is repeated, so the invariant
// survives concurrency instead of leaking a second probe.
//
// Selection and the in-flight increment happen under one lock, so two
// concurrent calls cannot both read the same pre-increment load.
func (b *Balancer) Pick(cands []Candidate, allows func(Candidate) bool, acquire func(Candidate) bool) (Candidate, PickStats, bool) {
	b.mu.Lock()
	defer b.mu.Unlock()

	var excluded map[string]struct{}
	for {
		first, second, stats := b.draw(cands, allows, excluded)
		if stats.Eligible == 0 {
			return Candidate{}, stats, false
		}

		winner := first
		if stats.Eligible > 1 {
			// The reservoir samples the PAIR uniformly but not its ORDER: the
			// second slot is always filled second, and a load tie resolves to
			// the first. Without this flip the second candidate never wins on
			// an idle fleet.
			if b.intn(2) == 1 {
				first, second = second, first
			}
			winner = first
			if b.inflight[second.Key()] < b.inflight[first.Key()] {
				winner = second
			}
		}

		if acquire != nil && !acquire(winner) {
			if excluded == nil {
				excluded = make(map[string]struct{}, 1)
			}
			excluded[winner.InstanceID] = struct{}{}
			continue
		}

		b.inflight[winner.Key()]++
		return winner, stats, true
	}
}

// draw runs a size-2 reservoir over the eligible candidates in a single pass —
// no intermediate slice, so the hot path allocates nothing.
func (b *Balancer) draw(cands []Candidate, allows func(Candidate) bool, excluded map[string]struct{}) (Candidate, Candidate, PickStats) {
	var first, second Candidate
	stats := PickStats{Total: len(cands)}

	for _, c := range cands {
		if c.Endpoint == "" {
			continue
		}
		stats.Addressed++
		if _, skip := excluded[c.InstanceID]; skip {
			continue
		}
		if allows != nil && !allows(c) {
			continue
		}
		stats.Eligible++
		switch stats.Eligible {
		case 1:
			first = c
		case 2:
			second = c
		default:
			switch b.intn(stats.Eligible) {
			case 0:
				first = c
			case 1:
				second = c
			}
		}
	}
	return first, second, stats
}

// Release frees the in-flight slot Pick reserved. The key is deleted at zero so
// a fleet churning through instance IDs cannot grow the map without bound.
func (b *Balancer) Release(c Candidate) {
	b.mu.Lock()
	defer b.mu.Unlock()

	k := c.Key()
	if n := b.inflight[k]; n > 1 {
		b.inflight[k] = n - 1
		return
	}
	delete(b.inflight, k)
}

// InFlight reports the reserved slot count for one instance.
func (b *Balancer) InFlight(c Candidate) int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.inflight[c.Key()]
}

// Tracked reports how many instances hold a reservation.
func (b *Balancer) Tracked() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return len(b.inflight)
}
